### PR TITLE
added adr for DB upgrade

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

added adr for DB upgrade

### CODE REVIEW

**Review by arnaudsjs — APPROVED** [review]

File: docs/adr/0000-database-upgrade-testing.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
+# Database Upgrade Testing
+
+* Status: accepted
+* Deciders: [Sander, Arnaud, Wouter]
+* Date: 12/01/2023
+
+## Context and Problem Statement
+
+When making updates to the database schema, it is important to test this.
```
Comment (sanderr) [inline docs/adr/0000-database-upgrade-testing.md:9]: I think we should expand a bit on this. We don't just want to test the migrations (we already do that), we want a uniform approach to set up any relevant preconditions for the tests. Since the considered options focus on that part, I think the problem statement should as well.

File: docs/adr/0000-database-upgrade-testing.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
+# Database Upgrade Testing
+
+* Status: accepted
+* Deciders: [Sander, Arnaud, Wouter]
+* Date: 12/01/2023
```
Comment (sanderr) [inline docs/adr/0000-database-upgrade-testing.md:5]: The template contains a date format, I think it could get confusing if we start using different ones across ADRs. I have no objection against changing the format in the template either if you prefer.
```suggestion
* Date: 2023-01-12
```

**Review by FloLey — APPROVED** [review]

File: docs/adr/0000-database-upgrade-testing.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+## Considered Options
+
+1. Write data corresponding to the old database schema directly in the testcase to the DB.
+2. Generate test data using old code and persist it.
+3. Dynamically check out old code to run the test
+
+## Decision Outcome
+
+We chose option 2 to test upgrades by
```
Comment (sanderr) [inline docs/adr/0000-database-upgrade-testing.md:24]: As we saw in @Hugo-Inmanta 's PR there are two flavors of this option:
a. we always do it like this
b. we are flexible and allow prepopulation in the test case iff it doesn't use any code from `inmanta`.  Directly accessing the database is always safe, since we're operating on a dump that reflects the old data and schema.

Option b could serve to limit pollution of the dump tool for very specific edge cases, and it might simplify the test by making them more directed (they put something in and make sure it comes out as expected, focusing only on that representative subset).
I would consider this to be the best of both worlds.

File: docs/adr/0000-database-upgrade-testing.md (hunk 1 above)
Comment (wouterdb) [reply docs/adr/0000-database-upgrade-testing.md:9]: I personally care much less about uniform and more about correct. 

Could you make a suggestion?

Comment (sanderr) [reply docs/adr/0000-database-upgrade-testing.md:9]: That makes sense. Either way, the part where some of us made mistakes against, is the precondition set up, not the testing itself. A suggestion:
```suggestion
When making updates to the database schema, it is important to test this. In case the migration includes any data
transformation, testing this commonly involves setting up some known preconditions, and verifying that they are
converted as expected when the migration script is executed. These preconditions should be set up in a way that
remains both valid and relevant for the test scenario, even when related code may change in the future.

How do we set up these preconditions so that the tests remain valid and relevant in the future?
```

And then perhaps a small addition to the decision drivers. I'll make a separate suggestion there.

File: docs/adr/0000-database-upgrade-testing.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+* Date: 2023-01-12
+
+## Context and Problem Statement
+
+When making updates to the database schema, it is important to test this.
+
+## Decision Drivers
+
+* We want to test that database update code correctly transforms data produced by older code to data that can be handled by the current code
+* But within the code base, this older code no longer exists
```
Comment (sanderr) [inline docs/adr/0000-database-upgrade-testing.md:14]:
```suggestion
* But within the code base, this older code no longer exists or has been modified, so it can not be relied upon in the test case before the migration, where database schema and code might be out of sync.
```
Part of https://github.com/inmanta/inmanta-core/pull/5362/files#r1069134210

File: docs/adr/0000-database-upgrade-testing.md (hunk 3 above)
Comment (wouterdb) [reply docs/adr/0000-database-upgrade-testing.md:29]: added

Comment (sanderr) [reply docs/adr/0000-database-upgrade-testing.md:29]: I don't see it. Forgot to push?

Comment (wouterdb) [reply docs/adr/0000-database-upgrade-testing.md]: :facepalm:

File: docs/adr/0000-database-upgrade-testing.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+1. Write data corresponding to the old database schema directly in the testcase to the DB.
+2. Generate test data using old code and persist it.
+3. Dynamically check out old code to run the test
+
+## Decision Outcome
+
+We chose option 1 and 2.
+
+1. When data is very simple or specific, write data to the database in the testcase, prior to the update
```
Comment (sanderr) [inline docs/adr/0000-database-upgrade-testing.md:31]:
```suggestion
1. When data is very simple or specific, write data to the database in the testcase, prior to the update. Make sure not to use any inmanta code.
```

**Review by sanderr — APPROVED** [review]

**Review by Hugo-Inmanta — APPROVED** [review]
